dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/AgentTooling.java

```
@@ -16,6 +20,26 @@
     registerWeakMapProvider();
   }
 
+  private static <K, V> Iterator<Provider> loadServiceProviders() {
+    return ServiceLoader.load(Provider.class, AgentInstaller.class.getClassLoader()).iterator();
+  }
+
+  public static <K, V> WeakCache<K, V> newWeakCache() {
+    Iterator<Provider> providers = loadServiceProviders();
+    if (providers.hasNext()) {
+      return providers.next().newWeakCache();
+    }
+    return new WeakCache.NoOpCache<>();
+  }
+
+  public static <K, V> WeakCache<K, V> newWeakCache(final long maxSize) {
+    Iterator<Provider> providers = loadServiceProviders();
+    if (providers.hasNext()) {
+      return providers.next().newWeakCache(maxSize);
+    }
+    return new WeakCache.NoOpCache<>();
```

Do we need the NoOp?  Or, would it better to just raise an IllegalStateException immediately?
I'm usually in favor of failing fast when possible.

---

it's always was failfast: https://github.com/DataDog/dd-trace-java/pull/1256/files#diff-c083f4bb26e5995615a99ef3bcd3f171R25 

But I will do it in less weird form